src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1724,104 +1724,300 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+                // tuple literal such as `(1, 2)`, `(null, null)`, `(x.P, y.M())`
+                return DeconstructTuple(targetPlaceholder, syntax, diagnostics, variables, deconstructionSteps, assignmentSteps);
+            }
+            else
+            {
+                return DeconstructWithDeconstruct(targetPlaceholder, syntax, diagnostics, variables, deconstructionSteps, assignmentSteps);
             }
-
-            return BindDeconstructWithDeconstruct(node, diagnostics, boundRHS, checkedVariables);
         }
 
-        private BoundExpression BindDeconstructWithTuple(AssignmentExpressionSyntax node, DiagnosticBag diagnostics, BoundExpression boundRHS, ImmutableArray<BoundExpression> checkedVariables)
+        /// <summary>
+        /// This will generate and stack appropriate deconstruction and assignment steps for a tuple type.
+        /// The produced deconstruction step has no Deconstruct method since the tuple already has distinct elements.
+        /// Returns false if there was an error.
+        /// </summary>
+        private bool DeconstructTuple(
+                        BoundDeconstructValuePlaceholder targetPlaceholder,
+                        AssignmentExpressionSyntax syntax,
+                        DiagnosticBag diagnostics,
+                        ImmutableArray<BoundExpression> variables,
+                        ArrayBuilder<BoundDeconstructionAssignmentOperator.DeconstructStep> deconstructionSteps,
+                        ArrayBuilder<BoundDeconstructionAssignmentOperator.AssignmentStep> assignmentSteps)
         {
-            // make a conversion for the tuple based on the LHS information (this also takes care of tuple literals without a natural type)
-            var lhsTypes = checkedVariables.SelectAsArray(v => v.Type);
-            TypeSymbol lhsAsTuple = TupleTypeSymbol.Create(locationOpt: null, elementTypes: lhsTypes, elementLocations: default(ImmutableArray<Location>), elementNames: default(ImmutableArray<string>), compilation: Compilation, diagnostics: diagnostics);
-            var typedRHS = GenerateConversionForAssignment(lhsAsTuple, boundRHS, diagnostics);
+            Debug.Assert(targetPlaceholder.Type.IsTupleType);
 
-            ImmutableArray<TypeSymbol> tupleTypes = typedRHS.Type.TupleElementTypes;
+            var tupleTypes = targetPlaceholder.Type.TupleElementTypes;
 
-            // figure out the pairwise conversions
-            var assignments = checkedVariables.SelectAsArray((variable, index, types) => MakeAssignmentInfo(variable, types[index], node, diagnostics), tupleTypes);
+            // Add step for deconstructing this target
+            var deconstructionStep = new BoundDeconstructionAssignmentOperator.DeconstructStep()
+            {
+                DeconstructMemberOpt = null,
+                TargetPlaceholder = targetPlaceholder,
+                OutputPlaceholders = tupleTypes.SelectAsArray(type => new BoundDeconstructValuePlaceholder(syntax, type))
```

Comment:
Consider using ```tupleTypes.SelectAsArray((t, s) => new BoundDeconstructValuePlaceholder(s, t), syntax)``` to avoid allocating a closure class. Similar comment for `DeconstructWithDeconstruct`.